Plutarch/List.hs

```diff
@@ -0,0 +1,189 @@
+
+import Plutarch
+import Plutarch.Bool
+import Plutarch.Builtin
+import Plutarch.Integer
+import Plutarch.Prelude
+import qualified PlutusCore as PLC
+
+--------------------------------------------------------------------------------
+
+data PList (a :: k -> Type) (s :: k)
+  = PCons (Term s a) (Term s (PList a))
+  | PNil
+
+-- Instances
+
+instance PIsData a => PlutusType (PList a) where
+  type PInner (PList a) b = PBuiltinList (PAsData a)
+  pcon' :: forall s. PList a s -> forall b. Term s (PInner (PList a) b)
+  pcon' (PCons x xs) = pconsBuiltin # (pdata x) # (pto xs)
+  pcon' PNil = pemptyList
+  pmatch' xs f =
+    pforce $ pif (pnullBuiltin # xs) (pdelay (f PNil)) (pdelay (f (PCons (pfromData $ pheadBuiltin # xs) (punsafeFrom $ ptailBuiltin # xs))))
+
+instance PEq a => PEq (PList a) where
+  (#==) xs ys =
+    punsafeBuiltin PLC.EqualsData
+      # (pmkList # punsafeCoerce xs)
+      # (pmkList # punsafeCoerce ys)
+
+--------------------------------------------------------------------------------
+-- Construction
+
+-- | Create a singleton list from an element
+psingleton :: PIsData a => Term s (a :--> PList a)
+psingleton = plam $ \x -> pcon (PCons x (pcon PNil))
+
+-- | The empty list
+pnil :: PIsData a => Term s (PList a)
+pnil = pcon PNil
+
+--------------------------------------------------------------------------------
+-- Querying
+
+-- | / O(n) /. Check if element is in the list
+pelem :: (PIsData a, PEq a) => Term s (a :--> PList a :--> PBool)
+pelem =
+  phoistAcyclic $
+    plam $ \needle ->
+      precList
+        (\self x xs -> pif (x #== needle) (pcon PTrue) (self # xs))
+        (\_self -> pcon PFalse)
+
+-- | / O(n) /. Count the number of elements in the list
+plength :: PIsData a => Term s (PList a :--> PInteger)
+plength =
+  plet
+    ( pfix #$ plam $ \self ls n ->
+        pmatch ls $ \case
+          PCons _x xs -> (self # xs # n + 1)
+          PNil -> n
+    )
+    $ \go -> plam $ \xs -> go # xs # 0
+
+pnull :: PIsData a => Term s (PList a :--> PBool)
+pnull =
+  plam $ \xs ->
+    pmatch xs $ \case
+      PCons _ _ -> pcon PFalse
+      PNil -> pcon PTrue
+    
+
+--------------------------------------------------------------------------------
+
+-- | Meta recursive list eliminator
+precList :: PIsData a => (Term s (PList a :--> r) -> Term s a -> Term s (PList a) -> Term s r) -> (Term s (PList a :--> r) -> Term s r) -> Term s (PList a :--> r)
+precList mcons mnil =
+  pfix #$ plam $ \self xs ->
+    pmatch xs $ \case
+      PCons x xs -> mcons self x xs
+      PNil -> mnil self
+
+-- | / O(n) /. Fold on a list right-associatively
+pfoldr :: PIsData a => Term s ((a :--> b :--> b) :--> b :--> PList a :--> b)
+pfoldr = phoistAcyclic $
+  plam $ \f z ->
+    precList
+      (\self x xs -> f # x # (self # xs))
+      (\_self -> z)
+
+punsafeHead :: PIsData a => Term s (PList a :--> a)
+punsafeHead =
+  phoistAcyclic $
+    plam $ \xs ->
+      pmatch xs $ \case
+        PCons x _ -> x
+        PNil -> perror
```

Comment:
We can use `HeadList` directly on this, as it is partial anyway? Same for `punsafeTail`.

Reply:
Maybe we should leave these out entirely then? Or make them return `PMaybe`s.